client/src/main/scala/org/http4s/client/DefaultClient.scala

```diff
@@ -0,0 +1,261 @@
+    *            is disposed when the returned task completes.  Attempts to read the
+    *            response body afterward will result in an error.
+    * @return The result of applying f to the response to req
+    */
+  def fetch[A](req: Request[F])(f: Response[F] => F[A]): F[A] =
+    run(req).use(f)
+
+  /** Submits a request, and provides a callback to process the response.
+    *
+    * @param req An effect of the request to submit
+    * @param f A callback for the response to req.  The underlying HTTP connection
+    *          is disposed when the returned task completes.  Attempts to read the
+    *          response body afterward will result in an error.
+    * @return The result of applying f to the response to req
+    */
+  def fetch[A](req: F[Request[F]])(f: Response[F] => F[A]): F[A] =
+    req.flatMap(fetch(_)(f))
+
+  /**
+    * Returns this client as a [[Kleisli]].  All connections created by this
+    * service are disposed on completion of callback task f.
+    *
+    * This method effectively reverses the arguments to `fetch`, and is
+    * preferred when an HTTP client is composed into a larger Kleisli function,
+    * or when a common response callback is used by many call sites.
+    */
+  def toKleisli[A](f: Response[F] => F[A]): Kleisli[F, Request[F], A] =
+    Kleisli(fetch(_)(f))
+
+  @deprecated("Use toKleisli", "0.18")
+  def toService[A](f: Response[F] => F[A]): Service[F, Request[F], A] =
+    toKleisli(f)
+
+  /**
+    * Returns this client as an [[HttpApp]].  It is the responsibility of
+    * callers of this service to run the response body to dispose of the
+    * underlying HTTP connection.
+    *
+    * This is intended for use in proxy servers.  `fetch`, `fetchAs`,
+    * [[toKleisli]], and [[streaming]] are safer alternatives, as their
+    * signatures guarantee disposal of the HTTP connection.
+    */
+  def toHttpApp: HttpApp[F] = Kleisli { req =>
+    /* Derived from https://github.com/typelevel/cats-effect/blob/c62c6c76b3066c500fca2f9e0897605bc12eb2a0/core/shared/src/main/scala/cats/effect/Resource.scala */
+
+    // Indirection for calling `loop` needed because `loop` must be @tailrec
+    def continue(
+        current: Resource[F, Any],
+        stack: List[Any => Resource[F, Any]],
+        release: F[Unit]): F[(Any, F[Unit])] =
+      loop(current, stack, release)
+
+    // Interpreter that knows how to evaluate a Resource data structure;
+    // Maintains its own stack for dealing with Bind chains
+    @tailrec def loop(
+        current: Resource[F, Any],
+        stack: List[Any => Resource[F, Any]],
+        release: F[Unit]): F[(Any, F[Unit])] =
+      current match {
+        case Resource.Allocate(resource) =>
+          F.bracketCase(resource) {
+            case (a, rel) =>
+              stack match {
+                case Nil => F.pure(a -> release.guarantee(rel(ExitCase.Completed)))
+                case f0 :: xs => continue(f0(a), xs, release.guarantee(rel(ExitCase.Completed)))
```

Comment:
We release in reverse order, so I think these guarantees are both backward?  We don't yet have a test that exposes this.